packages/compiler/src/core/deprecation.ts

```diff
@@ -0,0 +1,50 @@
+import { Program } from "./program.js";
+import { Type } from "./types.js";
+
+// TODO: No duplication!
+function createStateSymbol(name: string) {
+  return Symbol.for(`TypeSpec.${name}`);
+}
+
+const deprecatedKey = createStateSymbol("deprecated");
+
+/**
+ * Provides details on the deprecation of a given type.
+ */
+export interface DeprecationDetails {
+  /**
+   * The deprecation message to display when the type is used.
+   */
+  message: string;
+}
+
+/**
+ * Check if the given type is deprecated
+ * @param program Program
+ * @param type Type
+ */
+export function isDeprecated(program: Program, type: Type): boolean {
+  return program.stateMap(deprecatedKey).has(type);
```

Comment:
I think we can't use the stateMap for this otherwise it will carry over for child types.

Reply:
I was thinking that this is maybe something that shows up on the type itself but doesn't get cloned

Reply:
oh nvm, this works, the stateMap doesn't get copied over unless a decorator runs.

Reply:
Yeah, needs the decorator for the metadata to carry across with `is`, I'm explicitly not copying it across in the checker.